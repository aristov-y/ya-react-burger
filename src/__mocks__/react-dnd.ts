const mockDropRef = jest.fn();
const mockDragRef = jest.fn();

const useDrag = jest.fn().mockImplementation(() => [undefined, mockDragRef]);
const useDrop = jest.fn().mockImplementation(() => [undefined, mockDropRef]);

export {
  useDrag,
  useDrop
}
