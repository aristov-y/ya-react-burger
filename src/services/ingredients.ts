import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import fetchIngredients, { Ingredient } from '../utils/ingredients';

const ingredientsState = {
  ingredients: [] as Ingredient[],
  loading: false,
  hasError: false,
};

const loadIngredients = createAsyncThunk(
  'ingredients/load',
  async () => {
    return fetchIngredients();
  }
);

const ingredients = createSlice({
  name: 'ingredients',
  initialState: ingredientsState,
  reducers: {
    setIngredients: (state, action: PayloadAction<Ingredient[]>) => {
      state.ingredients = action.payload
    }
  },
  extraReducers: (builder) => {
    builder
      .addCase(loadIngredients.pending, (state) => {
        state.loading = true;
      })
      .addCase(loadIngredients.fulfilled, (state, action) => {
        state.ingredients = action.payload;
        state.loading = false;
        state.hasError = false;
      })
      .addCase(loadIngredients.rejected, (state, action) => {
        state.ingredients = [];
        state.loading = false;
        state.hasError = true;
      })
  }
})

const { setIngredients } = ingredients.actions;

export {
  ingredients,
  setIngredients,
  loadIngredients,
  ingredientsState
}
