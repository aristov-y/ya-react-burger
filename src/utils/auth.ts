import { deleteCookie, getCookie, setCookie } from './cookies';
import { TSimpleUserInfo } from '../types';

interface RegisterResponse {
  success: boolean,
  accessToken: string,
  "refreshToken": string,
  user: {
    email: string,
    name: string
  }
}

interface RefreshResponse {
  success: boolean,
  accessToken: string,
  refreshToken: string
}

interface UserResponse {
  success: boolean,
  user: {
    email: string,
    name: string
  }
}

interface SimpleResponse {
  success: boolean;
  message: string;
}

function responseParse(response: Response) {
  if (response.ok) {
    return response.json();
  }
  return response.json().then(err => Promise.reject(err));
}

function loginRequest(email: string, password: string): Promise<RegisterResponse> {
  return fetch('https://norma.nomoreparties.space/api/auth/login', {
    method: 'post',
    body: JSON.stringify({
      email,
      password
    }),
    headers: {
      'content-type': 'application/json'
    }
  })
    .then(responseParse)
    .then((data: RegisterResponse) => {
      if (data.success) {
        setCookie('token', data.accessToken);
        localStorage.setItem('token', data.refreshToken)
        return data;
      }
      return Promise.reject();
    });
}

function registerRequest(email: string, name: string, password: string): Promise<RegisterResponse> {
  return fetch('https://norma.nomoreparties.space/api/auth/register', {
    method: 'post',
    body: JSON.stringify({
      email,
      password,
      name
    }),
    headers: {
      'content-type': 'application/json'
    }
  })
    .then(responseParse)
    .then((data: RegisterResponse) => {
      if (data.success) {
        setCookie('token', data.accessToken);
        localStorage.setItem('token', data.refreshToken)
        return data;
      }
      return Promise.reject()
    })
}

function logoutRequest(): Promise<SimpleResponse> {
  const token = localStorage.getItem('token');
  return fetch('https://norma.nomoreparties.space/api/auth/logout', {
    method: 'post',
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify({
      token
    })
  })
    .then(responseParse)
    .then(data => {
      if (data.success) {
        deleteCookie('token');
        localStorage.removeItem('token');
      }
      return data;
    })
}

function tokenRequest(): Promise<RefreshResponse> {
  const token = localStorage.getItem('token');
  return fetch('https://norma.nomoreparties.space/api/auth/token', {
    method: 'post',
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify({
      token
    })
  })
    .then(responseParse)
    .then((data: RefreshResponse) => {
      if (data.success) {
        localStorage.setItem('token', data.refreshToken);
        return data;
      }
      return Promise.reject();
    })
}

function userInfoRequest(): Promise<UserResponse> {
  return fetch('https://norma.nomoreparties.space/api/auth/user', {
    method: 'get',
    headers: {
      'Authorization': getCookie('token') as string
    }
  })
    .then(responseParse)
    .then(data => {
      if (data.success) {
        return data;
      }
      return Promise.reject(data);
    })
    .catch(err => {
      if (err.message === 'jwt expired') {
        return tokenRequest()
          .then(data => {
            setCookie('token', data.accessToken);
            localStorage.setItem('token', data.refreshToken)
            return fetch('https://norma.nomoreparties.space/api/auth/user', {
              method: 'get',
              headers: {
                'Authorization': getCookie('token') as string
              }
            })
              .then(responseParse)
          })
      } else {
        return Promise.reject(err);
      }
    })
}

function updateUserInfo(user: TSimpleUserInfo): Promise<UserResponse> {
  return fetch('https://norma.nomoreparties.space/api/auth/user', {
    method: 'PATCH',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': getCookie('token') as string
    },
    body: JSON.stringify(user)
  })
    .then(responseParse)
    .then(data => {
      if (data.success) {
        return data;
      }
      return Promise.reject();
    })
}

function resetPasswordRequest(email: string): Promise<SimpleResponse> {
  return fetch('https://norma.nomoreparties.space/api/password-reset', {
    method: 'post',
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify({
      email
    })
  })
    .then(responseParse)
    .then(data => {
      if (data.success) {
        return data;
      }
      return Promise.reject(data);
    })
}

function resetPassword(token: string, password: string) {
  return fetch('https://norma.nomoreparties.space/api/password-reset/reset', {
    method: 'post',
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify({
      token,
      password
    })
  })
    .then(responseParse)
    .then(data => {
      if (data.success) {
        return data;
      }
      return Promise.reject(data);
    })
}

export {
  tokenRequest,
  loginRequest,
  logoutRequest,
  registerRequest,
  updateUserInfo,
  userInfoRequest,
  resetPasswordRequest,
  resetPassword
}
