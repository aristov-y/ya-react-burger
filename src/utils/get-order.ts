interface Order {
  name: string,
  order: {
    number: number
  },
  success: boolean
}

function getOrder(ingredientIds: string[]): Promise<Order> {
  return fetch('https://norma.nomoreparties.space/api/orders', {
    headers: {
      'Content-Type': 'application/json'
    },
    method: 'POST',
    body: JSON.stringify({ingredients: ingredientIds})
  })
    .then(response => {
      if (response.ok) {
        return response.json();
      }
      return Promise.reject(response.statusText)
    });
}

export default getOrder;
