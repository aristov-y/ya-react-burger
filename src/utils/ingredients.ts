interface Ingredient {
  _id: string;
  name: string;
  type: 'main' | 'bun' | 'sauce';
  proteins: number;
  fat: number;
  carbohydrates: number;
  calories: number;
  price: number;
  image: string;
  image_mobile: string;
  image_large: string;
  __v: number;
}

function fetchIngredients(): Promise<Ingredient[]> {
  return window.fetch('https://norma.nomoreparties.space/api/ingredients')
    .then(e => e.json())
    .then(e => e.data);
}

export type {
  Ingredient
}

export default fetchIngredients;
