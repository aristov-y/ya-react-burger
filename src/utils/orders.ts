import { getCookie } from './cookies';

interface FeedItem {
  createdAt: string
  ingredients: string[]
  name: string
  number: number
  status: string
  updatedAt: string
  _id: string
}

interface FeedResponse {
  success: boolean;
  orders: FeedItem[];
  total: number;
  totalToday: number;
}

function fetchFeed(): Promise<FeedResponse> {
  return fetch(`${process.env.REACT_APP_API_DOMAIN}/api/orders/all`)
    .then(response => {
      if (response.ok) {
        return response.json()
      }
      return Promise.reject(response.statusText)
    })
    .then(json => {
      if (json.success) {
        return json;
      }
      return Promise.reject()
    })
}

function fetchUserOrders(): Promise<FeedItem[]> {
  return fetch(`${process.env.REACT_APP_API_DOMAIN}/api/orders`, {
    headers: {
      'Authorization': getCookie('token') as string
    }
  })
    .then(response => {
      if (response.ok) {
        return response.json()
      }
      return Promise.reject(response.statusText)
    })
    .then(json => {
      if (json.success) {
        return json.orders;
      }
      return Promise.reject()
    })
}

export type {
  FeedItem
}

export {
  fetchFeed,
  fetchUserOrders
}
