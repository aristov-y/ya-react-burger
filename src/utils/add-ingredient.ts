import { Ingredient } from './ingredients';

function addIngredient(ingredients: Ingredient[], newIngredient: Ingredient) {
  if (newIngredient.type === 'bun') {
    return [
      newIngredient,
      ...ingredients.filter(e => e.type !== 'bun')
    ];
  } else {
    return [
      ...ingredients,
      newIngredient
    ];
  }
}

export default addIngredient;
