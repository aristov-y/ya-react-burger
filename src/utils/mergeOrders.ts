import { FeedItem } from './orders';

function mergeOrders(newOrders: FeedItem[], oldItems: FeedItem[]): FeedItem[] {
  const temp = [...oldItems];
  newOrders.forEach((val: FeedItem) => {
    const id = val._id;
    const idx = temp.findIndex(e => e._id === id);
    if (idx === -1) {
      temp.splice(0, 0, val);
    } else {
      temp.splice(idx, 1, { ...val });
    }
  });
  return temp;
}

export default mergeOrders;
