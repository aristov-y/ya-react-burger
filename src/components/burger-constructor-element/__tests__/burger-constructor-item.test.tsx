import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import BurgerConstructorItem from '../burger-constructor-item';
import { Ingredient } from '../../../utils/ingredients';
import { useDrag, useDrop } from 'react-dnd';

const mockDragRef = jest.fn();
const mockDropRef = jest.fn();
const mockUseDispatch = jest.fn();
const mockRemoveIngredient = jest.fn();
const mockReorderIngredient = jest.fn();

jest.mock('../../../services/store', () => ({
  __esModule: true,
  removeIngredient: function(...args: any[]) {
    return mockRemoveIngredient.call(this, ...args)
  },
  reorderIngredient: function(...args: any[]) { return mockReorderIngredient.call(this, ...args)}
}));
jest.mock('react-redux', () => ({
   useDispatch: function(...args: any[]) { return mockUseDispatch.call(this, ...args)}
}));

beforeEach(() => {
  (useDrag as jest.Mock).mockImplementation(() => [undefined, mockDragRef]);
  (useDrop as jest.Mock).mockImplementation(() => [undefined, mockDropRef]);
  mockUseDispatch.mockClear();
  mockRemoveIngredient.mockClear();
  mockReorderIngredient.mockClear();
});

const item: Ingredient = {
  _id: "1",
  type: 'main',
  price: 2,
  name: 'name',
  image: 'image',
  image_mobile: 'image_mobile',
  fat: 123,
  calories: 124,
  carbohydrates: 11,
  proteins: 0,
  image_large: 'image_large',
  __v: 1
}

describe('burger-constructor-item', function () {
  it('should render', function () {
    let component: ReactTestRenderer | null = null;
    act(() => {
      component = create(<BurgerConstructorItem item={item} index={2} />);
    });

    expect(component !== null).toBeTruthy();
    expect(component.toJSON()).toMatchSnapshot();

    expect(useDrag).toBeCalledTimes(1);
  });
});
