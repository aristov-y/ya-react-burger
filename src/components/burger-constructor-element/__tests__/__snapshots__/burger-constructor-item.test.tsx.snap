// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`burger-constructor-item should render 1`] = `
<div
  className="burger-constructor-item burger-constructor-item__draggable"
>
  DragIcon
  ConstructorElement
</div>
`;
